src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_agents.clj

```diff
@@ -43,10 +38,27 @@
 
 (declare send-flush-instance!)
 
+(schema/defn send-and-await-fill-with-poison-pills!
+  "Uses the modify-instance-agent to fill up the pool with poison pills
+
+  insert-fn should be a function that takes in a reference to the pool
+  and inserts some kind of poison pill
+
+  Blocks until the agent has completed filling the pool"
+  [pool-context :- jruby-schemas/PoolContext
+   insert-fn :- IFn]
+  (let [on-complete (promise)
+        pool-state (jruby-internal/get-pool-state pool-context)
+        fill-fn (fn []
+                  (jruby-internal/fill-pool-with-poison-pills! pool-state insert-fn)
+                  (deliver on-complete true))]
+    (send-agent (get-modify-instance-agent pool-context) fill-fn)
```

Comment:
Yeah definitely. Now that you've brought it up I hope we end up going with the way you talked about. Seems like fewer edge cases in that world